### ISSUE TITLE
Permormance Issues using RazorViewEngine-Partial

### ISSUE THREAD

**Issue by dwonisch** [NancyFx/Nancy#2167]
There seems to be a huge performance issue using many Partials on one Page.

I have created a [sample application](https://github.com/dwonisch/NancyPerformanceTest) to explain what I exactly mean.
The application has two views, both showing the same data. The plain view, using foreach loops, loads for about 5 ms. The view using partials is taking about 250 ms (even on Release build, still over 60 ms). (Measured with Firefox network tool).

We generated our controls in C# using IHtmlStrings before, which was lightning fast, but was too confusing for large parts. Reading HTML or Razor Syntax is much easier, so I would prefer to use partials instead.
After introducing partials in our application, the load time for the start page has increased from under 1 seconds to over 25 seconds, which would make this feature simply unusable for us.

I looked through the code up and down, and found out that a call to NancyRazorViewBase.Execute() for a simple partial without model [SaveButton.cshtml](https://gist.github.com/dwonisch/417089e5b83d2b0665bb), would take between 50ms and 70ms.
This is the place where everything is put together into the StringBuilder and I think the issue is that all parts are instanciated for each own. Putting it all together an then instanciate should provide a better performance, but I still have no clue how to solve this.

**Comment by dwonisch** [comment]
Another day, another update!

I could locate the issue, it is the huge amount of default ViewLocationConventions.
We are using our own conventions, 7 at the moment, addionally Nancy is adding 16 default conventions.

Because we are not using any localized views I can safely remove all of them.

This decreases loading time of the measured page from 739ms, to incredibly 67ms (measured in Firefox again).

If it would be okay, I will try to implement a caching strategy for known views, because the slow parts are in the DefaultViewLocator.GetCachedMatchinViews(string viewName) method.
This method is called for each convention and as expected for each view times(!) every known view, resulting in 95004 calls of DefaultViewLocator.NameMatchesView() method per Page Request.
By removing the default conventions I could reduce it to 11484 calls, still a lot but here is the point to optimize.

Still need some time to create the PR.

**Comment by dwonisch** [comment]
I think I would need some advice regarding ViewLocationConventions. Is there any purpose in using them? The only "useful" thing I found out is, that they are used to create an order in which the locations are scanned, if there is more than one view with the same name, in different locations.

Neither FileSystemLocationProvider nor ResourceViewLocationProvider is using this information when locating views. So all views should be known by it's names on start up anyway, but that might not be intended.

**Comment by khellang** [comment]
> Is there any purpose in using them?

What does this mean? Are you asking if they were added just for fun?

**Comment by phillip-haydon** [comment]
Our view locations are designed to mimic MVC's conventions to lower the barrier of entry. 

There's absolutely nothing wrong with defining your own conventions that suit your requirements. Every project I do with Nancy I define my own because I have 3 conventions and don't require any of the more complicated ones or the culture specific ones.

**Comment by phillip-haydon** [comment]
Also views cannot be known at startup because the names of views could differ based on module folders.

i.e you have HomeModule with an Index view, and a ProductModule with an Index view.

**Comment by grumpydev** [comment]
Does this still occur in release mode / with EnableRuntimeViewDiscovery set to false?

**Comment by dwonisch** [comment]
@khellang No, I'm not asking if they are added just for fun. I'm trying to understand what they are used for before applying any changes to them. In my project I can easily remove them without any side effects, except for better performance.
I have some failing unit tests that I'm analysing now, to better understand their behaviour.

@phillip-haydon I was talking about this method: https://github.com/NancyFx/Nancy/blob/45860c82e4df8e2d380997ddf1d19d61400fb145/src/Nancy/ViewEngines/DefaultViewLocator.cs#L172
I'm not totally sure, but I think this is called on startup or first request.

@grumpydev Yes, this does still occur with EnableRuntimeViewDiscovery set to false. Caching happens in DefaultViewLocator and this performance issue hase something to do with checking for conventions in DefaultViewResolver, before calling the locator.
